examples/semi_supervised/plot_self_training_varying_threshold.py

```diff
@@ -0,0 +1,106 @@
+"""
+=============================================
+Effect of varying threshold for self-training
+=============================================
+
+This example illustrates the effect of a varying threshold on self-training.
+The `breast_cancer` dataset is loaded, and labels are deleted such that only 50
+out of 569 samples have labels. A `SelfTrainingClassifier` is fitted on this
+dataset, with varying thresholds.
+
+The upper graph shows the amount of (initially unlabeled) samples that the
+classifier predicted labels for during training, and the accuracy of the
+classifier. The lower graph shows the last iteration in which a sample was
+labeled. All values are cross validated with 3 folds.
+
+At low thresholds (in [0.4, 0.5]), the classifier learns from samples that were
+labeled with a low confidence. These low-confidence samples are likely have
+incorrect predicted labels, and as a result, fitting on these incorrect labels
+produces a poor accuracy. Note that the classifier labels almost all of the
+samples, and only takes one iteration.
+
+For very high thresholds (in [0.9, 1)) we observe that the classifier does not
+augment its dataset (the amount of self-labeled samples is 0). As a result, the
+accuracy achieved with a threshold of 0.9999 is the same as a normal supervised
+classifier would achieve.
+
+The optimum accuracy lies in between both of these extremes at around 0.7.
+"""
+print(__doc__)
+
+# Authors: Oliver Rausch <rauscho@ethz.ch>
+# License: BSD
+
+import numpy as np
+import matplotlib.pyplot as plt
+from sklearn import datasets
+from sklearn.svm import SVC
+from sklearn.model_selection import StratifiedKFold
+from sklearn.semi_supervised import SelfTrainingClassifier
+from sklearn.metrics import accuracy_score
+from sklearn.utils import shuffle
+
+n_splits = 3
+
+X, y = datasets.load_breast_cancer(return_X_y=True)
+X, y = shuffle(X, y, random_state=42)
+y_true = y.copy()
+y[50:] = -1
+total_samples = y.shape[0]
+
+base_classifier = SVC(probability=True, gamma=0.001, random_state=42)
+
+x_values = np.arange(0.4, 1.05, 0.05)
+x_values = np.append(x_values, 0.99999)
+scores = np.empty((x_values.shape[0], n_splits))
+amount_labeled = np.empty((x_values.shape[0], n_splits))
+amount_iterations = np.empty((x_values.shape[0], n_splits))
+
+for (i, threshold) in enumerate(x_values):
+    self_training_clf = SelfTrainingClassifier(base_classifier,
+                                               threshold=threshold)
+
+    # We need manual cross validation so that we don't treat -1 as a separate
+    # class when computing accuracy
+    skfolds = StratifiedKFold(n_splits=n_splits, random_state=42)
```

Comment:
I don't think random_state does anything by default. This isn't shuffled StratifiedKFold is it?

Reply:
In a real world semi-supervised context, we would want something roughly like:

```py
class SemiSupervisedSplitter:
    """Meta-splitter which always includes unlabelled samples in training"""

    def __init__(self, cv):
        self.cv = cv

    def split(self, X, y, TOOD: handle more args):
        cv = check_cv(self.cv)
        labeled = y != -1
        labeled_idx = np.flatnonzero(labeled)
        unlabeled_idx = np.flatnonzero(~labeled)
        for train_idx, test_idx in cv(X[labeled_idx], y[labeled_idx]):
            yield np.concatenate([labeled_idx[train_idx], unlabeled_idx]), labeled_idx[test_idx]

    def get_n_splits(self, *args, **kwargs):
        return check_cv(self.cv).get_n_splits(*args, **kwargs)
```

should we be doing that here and demonstrating real world usage, even if it means smaller test set sizes? (Should we be adding such a meta-splitter to the library?)

Reply:
Why is that better than the cv splitter for cross_val_score or search cv?

I don't mind that approach (though y should not need to be provided). It would be possible in a pipeline with fit_resample (https://github.com/scikit-learn/scikit-learn/issues/3855#issuecomment-357949997)

Reply:
I'm waiting for someone to take up the fit_resample implementation, btw... Keen?

Reply:
Oh I wasn't aware that splitters can be passed to cross validation functions. In that case I much prefer your solution, and I do think that it would be a useful addition to the library for real world usage.

Reply:
> I'm waiting for someone to take up the fit_resample implementation, btw... Keen?

I'll have a look. Maybe after this PR is done

Reply:
Should I open a separate PR for the splitter? Then if that one is merged we can use it here?